File: csharp/Behavioral/19-State/State/AssignedState.cs
namespace State;

public class AssignedState : BaseJobState
{
    public override string Name => "Assigned";

    public override void Start(JobCard job)
    {
        job.StartedAt = DateTime.Now;
        job.TransitionTo(new InProgressState());
        Console.WriteLine($"  ▶️  Work started at {job.StartedAt:HH:mm:ss}");
    }

    // Reassignment is allowed while still Assigned
    public override void Assign(JobCard job, string handyman)
    {
        var previous = job.Assignee;
        job.Assignee = handyman;
        Console.WriteLine($"  🔄 Reassigned from {previous} to {handyman}");
    }
}

File: typescript/Behavioral/19-State/State/AssignedState.ts
import { BaseJobState } from "./BaseJobState";
import { InProgressState } from "./InProgressState";
import type { JobCard } from "../Context/JobCard";

export class AssignedState extends BaseJobState {
  readonly name = "Assigned";

  override start(job: JobCard): void {
    const startedAt = new Date();
    job.startedAt = startedAt;
    job.transitionTo(new InProgressState());
    console.log(`  ▶️  Work started at ${startedAt.toTimeString().slice(0, 8)}`);
  }

  // Reassignment is allowed while still Assigned
  override assign(job: JobCard, handyman: string): void {
    const previous = job.assignee;
    job.assignee = handyman;
    console.log(`  🔄 Reassigned from ${previous} to ${handyman}`);
  }
}
